File: coin-change/coin-change.py
import numpy as np
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        DP = [np.inf] * (amount + 1)
        
        DP[0] = 0
        for a in range(1, amount +1):
            for c in coins:
                if a - c >= 0:
                    DP[a] = min(DP[a], 1 + DP[a-c]) # update DP values
        return DP[amount] if DP[amount]!= np.inf else -1 

File: coin-change/coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // double inf = std::numeric_limits<double>::infinity();
        // vector<double> DP(amount+1, inf);
        int Max = amount + 1;
        vector<int> DP(amount + 1, Max);
        DP[0] = 0;
        for (int a = 1; a< amount+1; a++){
            for (int i = 0; i< coins.size(); i++){
                if (a >= coins[i]){
                    DP[a] = min(DP[a], 1 + DP[a-coins[i]]);
                }
            }
        }
        
        if (DP[amount] != amount + 1){
            return DP[amount];
        }else{
            return -1;
        }
    }
};
                 